pkg/controller/kubefedcluster/controller.go

```diff
@@ -98,43 +102,71 @@ func newClusterController(config *util.ControllerConfig, clusterHealthCheckConfi
 		&fedv1b1.KubeFedCluster{},
 		util.NoResyncPeriod,
 		&cache.ResourceEventHandlerFuncs{
-			DeleteFunc: cc.delFromClusterSet,
-			AddFunc:    cc.addToClusterSet,
+			DeleteFunc: func(obj interface{}) {
+				castObj, ok := obj.(*fedv1b1.KubeFedCluster)
+				if !ok {
+					tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+					if !ok {
+						klog.Errorf("Couldn't get object from tombstone %#v", obj)
+						return
+					}
+					castObj, ok = tombstone.Obj.(*fedv1b1.KubeFedCluster)
+					if !ok {
+						klog.Errorf("Tombstone contained object that is not expected %#v", obj)
+						return
+					}
+				}
+				cc.delFromClusterSet(castObj)
+			},
+			AddFunc: func(obj interface{}) {
+				castObj := obj.(*fedv1b1.KubeFedCluster)
+				cc.addToClusterSet(castObj)
+			},
 			UpdateFunc: func(oldObj, newObj interface{}) {
-				cc.delFromClusterSet(oldObj)
-				cc.addToClusterSet(newObj)
+				var clusterChanged bool
+				cluster := newObj.(*fedv1b1.KubeFedCluster)
+				cc.mu.Lock()
+				clusterData, ok := cc.clusterDataMap[cluster.Name]
+				if ok && !equality.Semantic.DeepEqual(clusterData.cachedObj.Spec, cluster.Spec) {
```

Comment:
Is the use of `metadata.generation` not sufficient to detect when non-status changes occur?